File: apps/api/app/core/sensitive_fields.py
from __future__ import annotations

import re
import unicodedata
from typing import Any


SENSITIVE_FIELD_KEY_PROTOCOL_VERSION = (
    "semantic_sensitive_field_key_segments_v1"
)
SENSITIVE_FIELD_SCAN_MAX_DEPTH = 128
SENSITIVE_FIELD_SCAN_MAX_ITEMS = 1_000_000

_SINGULAR_SEGMENTS = {
    "credentials": "credential",
    "keys": "key",
    "prompts": "prompt",
    "responses": "response",
    "secrets": "secret",
    "tokens": "token",
}
_SAFE_STATUS_TAILS = frozenset(
    {
        "available",
        "configured",
        "count",
        "counts",
        "digest",
        "enabled",
        "exposed",
        "fields",
        "hash",
        "present",
        "protocol",
        "redacted",
        "redaction",
        "required",
        "source",
        "status",
        "version",
    }
)
_SAFE_TOKEN_OPERATIONAL_SEGMENTS = frozenset(
    {
        "accounting",
        "audit",
        "batch",
        "budget",
        "cache",
        "chat",
        "chunk",
        "client",
        "consumed",
        "consumption",
        "creation",
        "concept",
        "content",
        "context",
        "conversation",
        "cost",
        "count",
        "counts",
        "end",
        "estimate",
        "estimated",
        "extraction",
        "fixed",
        "history",
        "hint",
        "input",
        "json",
        "label",
        "length",
        "limit",
        "local",
        "max",
        "maximum",
        "mid",
        "min",
        "minimum",
        "model",
        "mode",
        "neighbor",
        "original",
        "output",
        "overlap",
        "package",
        "per",
        "protocol",
        "quality",
        "question",
        "read",
        "remaining",
        "request",
        "selected",
        "size",
        "span",
        "start",
        "sufficiency",
        "task",
        "tokenizer",
        "total",
        "usage",
        "used",
        "version",
        "window",
        "write",
    }
)
_DANGEROUS_STORAGE_SEGMENTS = frozenset(
    {
        "archive",
        "backup",
        "blob",
        "bundle",
        "copy",
        "snapshot",
        "value",
    }
)
_PUBLIC_PRIVATE_COMPACT_KEYS = frozenset(
    {
        "promptpack",
    }
)
_COMPACT_DANGEROUS_PATTERNS = (
    re.compile(r"(?:^|.*)apikey(?:backup|blob|bundle|copy|secret|value)?$"),
    re.compile(
        r"(?:^|.*)(?:access|admin|api|auth|bearer|client|id|oauth|refresh|"
        r"session|signing)?token(?:archive|backup|blob|bundle|copy|key|secret|"
        r"value)?$"
    ),
    re.compile(r"(?:^|.*)credential(?:archive|backup|blob|bundle|copy|s)?$"),
    re.compile(
        r"(?:^|.*)(?:providerrawresponse|rawproviderresponse|"
        r"providerresponse|providerpayload|rawproviderpayload)"
        r"(?:archive|backup|blob|bundle|copy|snapshot|value)?$"
    ),
    re.compile(
        r"(?:^|.*)system(?:content|prompt)"
        r"(?:archive|backup|blob|bundle|copy|snapshot|value)?$"
    ),
)


def semantic_key_segments(value: object) -> tuple[str, ...]:
    """Return deterministic semantic segments for case/separator variants."""

    text = unicodedata.normalize("NFKC", str(value).strip())
    text = re.sub(r"(?<=[A-Z])(?=[A-Z][a-z])", "_", text)
    text = re.sub(r"(?<=[a-z0-9])(?=[A-Z])", "_", text)
    raw_segments = re.findall(r"[a-z0-9]+", text.casefold())
    return tuple(_SINGULAR_SEGMENTS.get(item, item) for item in raw_segments)


def _has_contiguous_segments(
    segments: tuple[str, ...],
    expected: tuple[str, ...],
) -> bool:
    width = len(expected)
    return any(
        segments[index : index + width] == expected
        for index in range(0, len(segments) - width + 1)
    )


def _safe_observability_key(segments: tuple[str, ...]) -> bool:
    if not segments:
        return False
    if segments[0] == "has":
        return True
    if any(item in {"expose", "exposed", "exposes"} for item in segments):
        return True
    return segments[-1] in _SAFE_STATUS_TAILS


def _safe_token_operational_key(segments: tuple[str, ...]) -> bool:
    if "token" not in segments or len(segments) <= 1:
        return False
    non_token = tuple(item for item in segments if item != "token")
    return bool(non_token) and all(
        item in _SAFE_TOKEN_OPERATIONAL_SEGMENTS for item in non_token
    )


def sensitive_field_key_reason(
    value: object,
    *,
    include_public_private: bool = False,
) -> str | None:
    """Classify structural secret/provider/prompt field names.

    The protocol deliberately distinguishes credential semantics from ordinary
    token accounting.  For example, ``token_count`` and
    ``context_package_token_budget`` are operational metrics, while
    ``auth_token`` and ``tokenBackup`` are credentials.
    """

    segments = semantic_key_segments(value)
    if not segments:
        return None
    compact = "".join(segments)
    if include_public_private and compact in _PUBLIC_PRIVATE_COMPACT_KEYS:
        return "private_profile_prompt_container"

    has_profile_document = _has_contiguous_segments(
        segments, ("profile", "json")
    ) or compact.startswith("profilejson")
    has_api_key = _has_contiguous_segments(
        segments, ("api", "key")
    ) or "apikey" in compact
    has_provider_response = (
        (
            "provider" in segments
            and (
                "response" in segments
                or "payload" in segments
            )
        )
        or "providerrawresponse" in compact
        or "rawproviderresponse" in compact
        or "providerresponse" in compact
        or "providerpayload" in compact
        or "rawproviderpayload" in compact
    )
    has_system_prompt = (
        "system" in segments
        and (
            "prompt" in segments
            or "content" in segments
        )
    ) or "systemprompt" in compact or "systemcontent" in compact
    has_credential = "credential" in segments or "credential" in compact
    has_authorization = (
        "auth" in segments
        or "authorization" in segments
        or compact in {"auth", "authorization", "authorizationheader"}
        or compact.startswith("authtoken")
    )
    has_secret = (
        "password" in segments
        or "secret" in segments
        or compact.endswith("password")
        or compact.endswith("secret")
    )
    has_token = "token" in segments or any(
        pattern.fullmatch(compact)
        for pattern in _COMPACT_DANGEROUS_PATTERNS[1:2]
    )
    has_dangerous_storage_semantics = bool(
        _DANGEROUS_STORAGE_SEGMENTS.intersection(segments)
    )

    if (
        has_api_key
        or has_provider_response
        or has_credential
        or has_authorization
        or has_secret
        or has_system_prompt
    ) and not has_dangerous_storage_semantics and _safe_observability_key(
        segments
    ):
        return None
    if has_token and _safe_token_operational_key(segments):
        return None

    if has_profile_document:
        return "private_profile_document"
    if has_api_key:
        return "api_key"
    if has_provider_response:
        return "provider_response"
    if has_system_prompt:
        return "system_prompt"
    if has_credential:
        return "credential"
    if has_authorization:
        return "authorization"
    if has_secret:
        return "secret"
    if has_token:
        return "token"
    if any(pattern.fullmatch(compact) for pattern in _COMPACT_DANGEROUS_PATTERNS):
        return "dangerous_compound"
    return None


def _safe_sensitive_observability_value(key: str, value: Any) -> bool:
    """Allow only typed, content-free forms of otherwise sensitive metrics."""

    segments = semantic_key_segments(key)
    if segments == ("provider", "response", "persisted"):
        return value is False
    if segments == ("cacheable", "system", "prompt", "sha256"):
        return bool(
            isinstance(value, str)
            and re.fullmatch(r"[0-9a-f]{64}", value)
        )
    if segments in {
        ("cacheable", "system", "prompt", "utf8", "byte"),
        ("cacheable", "system", "prompt", "utf8", "bytes"),
    }:
        return type(value) is int and value >= 0
    return False


def sensitive_field_paths(
    value: Any,
    *,
    include_public_private: bool = False,
) -> tuple[str, ...]:
    """Return bounded, deterministic paths without reading or echoing values."""

    paths: list[str] = []
    stack: list[tuple[Any, tuple[str, ...], int]] = [(value, (), 0)]
    visited = 0
    while stack:
        current, path, depth = stack.pop()
        visited += 1
        if visited > SENSITIVE_FIELD_SCAN_MAX_ITEMS:
            raise ValueError("sensitive field scan exceeds its item bound")
        if depth > SENSITIVE_FIELD_SCAN_MAX_DEPTH:
            raise ValueError("sensitive field scan exceeds its depth bound")
        model_dump = getattr(current, "model_dump", None)
        if callable(model_dump):
            current = model_dump()
        if isinstance(current, dict):
            for raw_key, child in current.items():
                key = str(raw_key)
                child_path = (*path, key)
                if sensitive_field_key_reason(
                    key,
                    include_public_private=include_public_private,
                ) and not _safe_sensitive_observability_value(key, child):
                    paths.append(".".join(child_path))
                stack.append((child, child_path, depth + 1))
        elif isinstance(current, (list, tuple)):
            for index, child in enumerate(current):
                stack.append(
                    (child, (*path, f"[{index}]"), depth + 1)
                )
    return tuple(sorted(set(paths)))

File: apps/web/src/lib/sensitive-fields.ts
export const SENSITIVE_FIELD_KEY_PROTOCOL_VERSION =
  "semantic_sensitive_field_key_segments_v1" as const;

export type SensitiveFieldKeyReason =
  | "private_profile_prompt_container"
  | "private_profile_document"
  | "api_key"
  | "provider_response"
  | "system_prompt"
  | "credential"
  | "authorization"
  | "secret"
  | "token"
  | "dangerous_compound";

export type PESensitiveKeyKind =
  | "provider_raw_response"
  | "credentials";

const singularSegments = new Map([
  ["credentials", "credential"],
  ["keys", "key"],
  ["prompts", "prompt"],
  ["responses", "response"],
  ["secrets", "secret"],
  ["tokens", "token"],
]);

const safeStatusTails = new Set([
  "available",
  "configured",
  "count",
  "counts",
  "digest",
  "enabled",
  "exposed",
  "fields",
  "hash",
  "present",
  "protocol",
  "redacted",
  "redaction",
  "required",
  "source",
  "status",
  "version",
]);

const safeTokenOperationalSegments = new Set([
  "audit",
  "batch",
  "budget",
  "cache",
  "chunk",
  "client",
  "consumed",
  "consumption",
  "concept",
  "content",
  "context",
  "conversation",
  "cost",
  "count",
  "counts",
  "end",
  "estimate",
  "estimated",
  "extraction",
  "fixed",
  "history",
  "hint",
  "input",
  "label",
  "length",
  "limit",
  "local",
  "max",
  "maximum",
  "mid",
  "min",
  "minimum",
  "model",
  "neighbor",
  "original",
  "output",
  "overlap",
  "package",
  "per",
  "protocol",
  "quality",
  "question",
  "remaining",
  "request",
  "selected",
  "size",
  "span",
  "start",
  "sufficiency",
  "task",
  "tokenizer",
  "total",
  "usage",
  "used",
  "version",
  "window",
  "write",
]);

const dangerousStorageSegments = new Set([
  "archive",
  "backup",
  "blob",
  "bundle",
  "copy",
  "snapshot",
  "value",
]);

const publicPrivateCompactKeys = new Set(["promptpack"]);

const compactDangerousPatterns = [
  /^.*apikey(?:backup|blob|bundle|copy|secret|value)?$/,
  /^.*(?:access|admin|api|auth|bearer|client|id|oauth|refresh|session|signing)?token(?:archive|backup|blob|bundle|copy|key|secret|value)?$/,
  /^.*credential(?:archive|backup|blob|bundle|copy|s)?$/,
  /^.*(?:providerrawresponse|rawproviderresponse|providerresponse|providerpayload|rawproviderpayload)(?:archive|backup|blob|bundle|copy|snapshot|value)?$/,
  /^.*system(?:content|prompt)(?:archive|backup|blob|bundle|copy|snapshot|value)?$/,
];

function pythonCasefoldForAsciiSegments(value: string): string {
  // These are the only NFKC-stable Unicode case-fold expansions that change
  // the ASCII alphanumeric segments produced by Python's str.casefold().
  return value
    .replace(/[\u00df\u1e9e]/g, "ss")
    .replace(/\u01f0/g, "j\u030c")
    .replace(/\u1e96/g, "h\u0331")
    .replace(/\u1e97/g, "t\u0308")
    .replace(/\u1e98/g, "w\u030a")
    .replace(/\u1e99/g, "y\u030a")
    .toLowerCase();
}

export function semanticSensitiveKeySegments(value: unknown): string[] {
  const camelSegmented = String(value)
    .trim()
    .normalize("NFKC")
    .replace(/([A-Z]+)([A-Z][a-z])/g, "$1_$2")
    .replace(/([a-z0-9])([A-Z])/g, "$1_$2");
  const rawSegments =
    pythonCasefoldForAsciiSegments(camelSegmented).match(/[a-z0-9]+/g) ?? [];
  return rawSegments.map(
    (segment) => singularSegments.get(segment) ?? segment,
  );
}

function hasContiguousSegments(
  segments: string[],
  expected: string[],
): boolean {
  for (
    let index = 0;
    index <= segments.length - expected.length;
    index += 1
  ) {
    if (
      expected.every(
        (segment, offset) => segments[index + offset] === segment,
      )
    ) {
      return true;
    }
  }
  return false;
}

function safeObservabilityKey(segments: string[]): boolean {
  if (segments.length === 0) return false;
  if (segments[0] === "has") return true;
  if (
    segments.some((segment) =>
      ["expose", "exposed", "exposes"].includes(segment),
    )
  ) {
    return true;
  }
  return safeStatusTails.has(segments[segments.length - 1]);
}

function safeTokenOperationalKey(segments: string[]): boolean {
  if (!segments.includes("token") || segments.length <= 1) return false;
  const nonToken = segments.filter((segment) => segment !== "token");
  return (
    nonToken.length > 0 &&
    nonToken.every((segment) => safeTokenOperationalSegments.has(segment))
  );
}

export function sensitiveFieldKeyReason(
  value: unknown,
  options: { includePublicPrivate?: boolean } = {},
): SensitiveFieldKeyReason | null {
  const segments = semanticSensitiveKeySegments(value);
  if (segments.length === 0) return null;
  const compact = segments.join("");
  if (
    options.includePublicPrivate &&
    publicPrivateCompactKeys.has(compact)
  ) {
    return "private_profile_prompt_container";
  }

  const hasProfileDocument =
    hasContiguousSegments(segments, ["profile", "json"]) ||
    compact.startsWith("profilejson");
  const hasApiKey =
    hasContiguousSegments(segments, ["api", "key"]) ||
    compact.includes("apikey");
  const hasProviderResponse =
    (segments.includes("provider") &&
      (segments.includes("response") || segments.includes("payload"))) ||
    compact.includes("providerrawresponse") ||
    compact.includes("rawproviderresponse") ||
    compact.includes("providerresponse") ||
    compact.includes("providerpayload") ||
    compact.includes("rawproviderpayload");
  const hasSystemPrompt =
    (segments.includes("system") &&
      (segments.includes("prompt") || segments.includes("content"))) ||
    compact.includes("systemprompt") ||
    compact.includes("systemcontent");
  const hasCredential =
    segments.includes("credential") || compact.includes("credential");
  const hasAuthorization =
    segments.includes("auth") ||
    segments.includes("authorization") ||
    ["auth", "authorization", "authorizationheader"].includes(compact) ||
    compact.startsWith("authtoken");
  const hasSecret =
    segments.includes("password") ||
    segments.includes("secret") ||
    compact.endsWith("password") ||
    compact.endsWith("secret");
  const hasToken =
    segments.includes("token") || compactDangerousPatterns[1].test(compact);
  const hasDangerousStorageSemantics = segments.some((segment) =>
    dangerousStorageSegments.has(segment),
  );

  if (
    (hasApiKey ||
      hasProviderResponse ||
      hasCredential ||
      hasAuthorization ||
      hasSecret ||
      hasSystemPrompt) &&
    !hasDangerousStorageSemantics &&
    safeObservabilityKey(segments)
  ) {
    return null;
  }
  if (hasToken && safeTokenOperationalKey(segments)) return null;

  if (hasProfileDocument) return "private_profile_document";
  if (hasApiKey) return "api_key";
  if (hasProviderResponse) return "provider_response";
  if (hasSystemPrompt) return "system_prompt";
  if (hasCredential) return "credential";
  if (hasAuthorization) return "authorization";
  if (hasSecret) return "secret";
  if (hasToken) return "token";
  if (compactDangerousPatterns.some((pattern) => pattern.test(compact))) {
    return "dangerous_compound";
  }
  return null;
}

export function peSensitiveKeyKind(
  value: unknown,
): PESensitiveKeyKind | null {
  const segments = semanticSensitiveKeySegments(value);
  const reason = sensitiveFieldKeyReason(value, {
    includePublicPrivate: true,
  });
  if (reason === "provider_response") return "provider_raw_response";
  if (reason !== null) return "credentials";

  const dangerousStorage = segments.some((segment) =>
    dangerousStorageSegments.has(segment),
  );
  for (const [left, right] of [
    ["raw", "output"],
    ["raw", "response"],
  ]) {
    for (let index = 0; index < segments.length - 1; index += 1) {
      if (segments[index] !== left || segments[index + 1] !== right) {
        continue;
      }
      if (index + 2 === segments.length || dangerousStorage) {
        return "provider_raw_response";
      }
    }
  }

  const lastHeaderIndex = segments.reduce(
    (lastIndex, segment, index) =>
      ["header", "headers"].includes(segment) ? index : lastIndex,
    -1,
  );
  if (
    lastHeaderIndex >= 0 &&
    (lastHeaderIndex === segments.length - 1 || dangerousStorage)
  ) {
    return "credentials";
  }
  return null;
}
